### PR TITLE
feat: git fetcher with one repo and creates cache

### PR DESCRIPTION
**Pull request by Slug-Boi**
This branch allows the user to fetch from a single repo using the -fetch flag. The user can specify what files or paths in the repo to fetch from using a specfile as a whitelist. This also uses the depth modifier to ensure only the latest commit in a repo is used making it a bit more performant. 

Resolves [#39](https://github.com/orgs/NovoNordisk-OpenSource/projects/2/views/3?pane=issue&itemId=56773537)
Resolves [#41](https://github.com/orgs/NovoNordisk-OpenSource/projects/2/views/3?pane=issue&itemId=56772812)

### CODE REVIEW

**Comment by Slug-Boi** [comment]
The tests will currently fail as the fix to these is on the chore: restructure branch :shrug:

File: src/Fetcher/fetcher.go
Comment (Haruyzal) [inline src/Fetcher/fetcher.go:64]: Think I'm still confused as to why we have to create a dummy repo? ;-;

File: src/Fetcher/fetcher_test.go
Comment (Haruyzal) [inline src/Fetcher/fetcher_test.go:45]: +1 on this TODO to ensure the CSV file is valid and its content is formatted as expected :)

**Review by Haruyzal — CHANGES_REQUESTED** [review]

**Review by Alex-itu — APPROVED** [review]
Looks fine, as long as we merge feat_git_fetcher_mutiple branch after this

File: ci/go.sum
Comment (zqueamish) [inline ci/go.sum:1]: Out of pure curiosity, how come /gocarina is no longer needed?

File: src/HTML/html_test.go
Comment (Haruyzal) [inline src/HTML/html_test.go:1]: I see that we're changing this file name, and as such I'd recommend that we first merge `docs_basic_merger` -> `feat_basic_merger` -> `main` first, to avoid redoing this namechange :)

File: src/Fetcher/fetcher_test.go
Comment (Slug-Boi) [reply src/Fetcher/fetcher_test.go:45]: This can be done will do it later today :)

File: ci/go.sum
Comment (Slug-Boi) [reply ci/go.sum:1]: Good question. Idk xD I'm guessing it may have been a package we used at one point and then deleted from imports thus it being gone. Either way should be fine. The go.sum file is only used by your own compiler and is even generated as you run go mod tidy and such so shouldn't matter too much :0

File: src/Fetcher/fetcher.go
Comment (Slug-Boi) [reply src/Fetcher/fetcher.go:24]: This is not an error rather a mode specifier. Certain OS methods pertaining to creation, reading and writing of files need a mode of operation for perms to be specified (how much privilege does the program have and such things) 
you can calculate the perm level on websites like this https://chmod-calculator.com/

**Review by JonasSkjodt — APPROVED** [review]

File: src/Fetcher/fetcher.go
Comment (Slug-Boi) [reply src/Fetcher/fetcher.go:64]: I'm assuming you mean as to why we create a dummy repo (which is maybe also just bad wording on our end tbf) It just initializes a new git repo with nothing set. This allows us to manipulate the repo using the git CLI commands that we all know and love. Thus we can active sparse checkout and set a whitelist before ever seeing the actual repo we would want  to pull from. This ensures we setup the rules correctly to sparsely and selectively collect the files

File: src/HTML/html_test.go
Comment (Slug-Boi) [reply src/HTML/html_test.go:1]: I can add this as a line in the description of this PR nicely spotted

File: ci/go.sum
Comment (zqueamish) [reply ci/go.sum:1]: Ah, alrighty c: thank you, and good to know!

**Comment by Slug-Boi** [comment]
Tests are green

File: src/HTML/html_test.go
Comment (Haruyzal) [reply src/HTML/html_test.go:1]: @Slug-Boi - this is no longer relevant due to the idiomatic folder structure changes made and pushed earlier, causing August's branch to adhere to these changes as well as the renames c:

Relevant PR: https://github.com/Agile-Arch-Angels/decentralized-tech-radar_dev/pull/15

File: src/Fetcher/fetcher.go
Comment (Haruyzal) [reply src/Fetcher/fetcher.go:64]: Ahh okay, that makes sense! We can write your explanation down in the fetcher's docs? I found that it made a lot of sense at least!

File: src/Fetcher/fetcher_test.go
Comment (Haruyzal) [reply src/Fetcher/fetcher_test.go:45]: Awesome! Looking forward to seeing those sweet green tests :')

Comment (Slug-Boi) [reply src/Fetcher/fetcher_test.go]: @Haruyzal This is now complete feel free to review the new addition to the test and approve if it looks good 👍

File: src/Fetcher/fetcher.go
Comment (August-Brandt) [inline src/Fetcher/fetcher.go:107]: Can we use more descriptive variable names here instead of `s` and `d`?

File: src/Fetcher/fetcher_test.go
Comment (August-Brandt) [inline src/Fetcher/fetcher_test.go:68]: `defer os.Remove(<file>)` should be called right after the file is created. Not at be bottom of the function

**Review by August-Brandt — CHANGES_REQUESTED** [review]
I have some changes i would like to be done before merge

**Review by Haruyzal — APPROVED** [review]
The new test looks good! Great job! Please do merge after looking at August's req :)

File: src/Fetcher/fetcher.go
Comment (Slug-Boi) [reply src/Fetcher/fetcher.go:64]: Feel free to add it to the fetcher docs yes

File: src/Fetcher/fetcher_test.go
Comment (Slug-Boi) [reply src/Fetcher/fetcher_test.go:68]: Defer functions are called any time a program or function closes so this doesn't matter. If anything it should just be moved to the top of a function as per usual go standard 🤷

File: src/Fetcher/fetcher.go
Comment (Haruyzal) [reply src/Fetcher/fetcher.go:64]: Sure thing! Will be done on a separate branch after the multi-fetcher has been approved :)

File: src/Fetcher/fetcher_test.go
Comment (Slug-Boi) [reply src/Fetcher/fetcher_test.go]: I have moved them to the top of the test

File: src/Fetcher/fetcher.go
Comment (Slug-Boi) [reply src/Fetcher/fetcher.go]: I have changed the names to be more descriptive

**Review by August-Brandt — APPROVED** [review]
Looks good now

**Pull request merged**
